src/main/java/rx/observables/AsyncOnSubscribe.java

```diff
@@ -0,0 +1,510 @@
+            this(null, next, onUnsubscribe);
+        }
+
+        public AsyncOnSubscribeImpl(Func3<S, Long, Observer<Observable<? extends T>>, S> nextFunc) {
+            this(null, nextFunc, null);
+        }
+
+        @Override
+        protected S generateState() {
+            return generator == null ? null : generator.call();
+        }
+
+        @Override
+        protected S next(S state, long requested, Observer<Observable<? extends T>> observer) {
+            return next.call(state, requested, observer);
+        }
+
+        @Override
+        protected void onUnsubscribe(S state) {
+            if (onUnsubscribe != null)
+                onUnsubscribe.call(state);
+        }
+    }
+
+    @Override
+    public final void call(Subscriber<? super T> actualSubscriber) {
+        S state = generateState();
+        UnicastSubject<Observable<T>> subject = UnicastSubject.<Observable<T>> create();
+        AsyncOuterSubscriber<S, T> outerSubscriberProducer = new AsyncOuterSubscriber<S, T>(this, state, subject);
+        actualSubscriber.add(outerSubscriberProducer);
+        Observable.concat(subject).unsafeSubscribe(Subscribers.wrap(actualSubscriber));
+        actualSubscriber.setProducer(outerSubscriberProducer);
+    }
+
+    private static class AsyncOuterSubscriber<S, T> extends ConcurrentLinkedQueue<Long>implements Producer, Subscription, Observer<Observable<? extends T>> {
+        /** */
+        private static final long serialVersionUID = -7884904861928856832L;
+
+        private volatile int isUnsubscribed;
+        @SuppressWarnings("rawtypes")
+        private static final AtomicIntegerFieldUpdater<AsyncOuterSubscriber> IS_UNSUBSCRIBED = AtomicIntegerFieldUpdater.newUpdater(AsyncOuterSubscriber.class, "isUnsubscribed");
+
+        private final AsyncOnSubscribe<S, T> parent;
+        private final SerializedObserver<Observable<? extends T>> serializedSubscriber;
+        private final Set<Subscription> subscriptions = new HashSet<Subscription>();
+
+        private boolean hasTerminated = false;
+        private boolean onNextCalled = false;
+
+        private S state;
+
+        private final UnicastSubject<Observable<T>> merger;
+
+        public AsyncOuterSubscriber(AsyncOnSubscribe<S, T> parent, S initialState, UnicastSubject<Observable<T>> merger) {
+            this.parent = parent;
+            this.serializedSubscriber = new SerializedObserver<Observable<? extends T>>(this);
+            this.state = initialState;
+            this.merger = merger;
+        }
+
+        @Override
+        public void unsubscribe() {
+            if (IS_UNSUBSCRIBED.compareAndSet(this, 0, 1)) {
+                // it's safe to process terminal behavior
+                if (isEmpty()) {
+                    parent.onUnsubscribe(state);
+                }
+                for (Subscription s : subscriptions) {
```

Comment:
If you could write a test that demonstrates your theory then that would help to debug/refine.